runtime/memprof.c

```diff
@@ -0,0 +1,316 @@
+  y = y ^ ((y << 15) & 0xefc60000);
+  y = y ^ (y >> 18);
+
+  mt_index++;
+  return y*2.3283064365386962890625e-10 + /* 2^-32 */
+          1.16415321826934814453125e-10; /* 2^-33 */
+}
+
+/* Microsoft's C compiler does not define M_PI by default. */
+static const double PI = 3.141592653589793;
+
+/* C99's lgammaf function is not available in some compilers
+   (including MSVC). Here is our own approximate implementation of the
+   log-factorial function.
+
+   Requirement : n is a non-negative integer.
+
+   We use Ramanujan's formula. For n < 10^8, the absolute error is
+   less than 10^-6, which is way better than what we need.
+ */
+static double lfact(double n) {
+  static double tab[10] = {
+    0.0000000000, 0.0000000000, 0.6931471806, 1.7917594692, 3.1780538303,
+    4.7874917428, 6.5792512120, 8.5251613611, 10.6046029027, 12.8018274801 };
+  if(n < 10)
+    return tab[(int)n];
+  return n*(log(n) - 1) + (1./6)*log(((8*n + 4)*n + 1)*n + 1./34)
+       + 0.5723649429; /* log(pi)/2 */
+}
+
+static double next_mt_generate_poisson;
+/* Max returned value : 2^30-2. Assumes lambda >= 0. */
+static int32_t mt_generate_poisson(double len) {
+  double cur_lambda = lambda * len;
+  CAMLassert(cur_lambda >= 0 && cur_lambda < 1e20);
+
+  if(caml_memprof_suspended || cur_lambda == 0)
+    return 0;
+
+  if(cur_lambda < 20) {
+    /* First algorithm when [cur_lambda] is small: we proceed by
+       repeated simulations of exponential distributions. */
+
+    next_mt_generate_poisson -= cur_lambda;
+    if(next_mt_generate_poisson > 0) {
+      /* Fast path if [cur_lambda] is small: we reuse the same
+         exponential sample accross several calls to
+         [mt_generate_poisson]. */
+      return 0;
+    } else {
+      /* We use the float versions of exp/log, since these functions
+         are significantly faster, and we really don't need much
+         precision here. The entropy contained in
+         [next_mt_generate_poisson] is anyway bounded by the entropy
+         provided by [mt_generate_uniform], which is 32bits. */
+      double p = expf(-next_mt_generate_poisson);
+      int32_t k = 0;
+      do {
+        k++;
+        p *= mt_generate_uniform();
+      } while(p > 1);
+
+      /* [p] is now uniformly distributed in [0, 1] and independent
+         from other variables (including [k]). We can therefore reuse
+         [p] for reinitializing [next_mt_generate_poisson]. */
+      next_mt_generate_poisson = -logf(p);
+
+      return k;
+    }
+
+  } else {
+
+    /* Second algorithm when [cur_lambda] is large. Taken from: */
+    /* The Computer Generation of Poisson Random Variables
+       A. C. Atkinson Journal of the Royal Statistical Society.
+       Series C (Applied Statistics) Vol. 28, No. 1 (1979), pp. 29-35
+       "Method PA" */
+
+    double c, beta_rec, k, log_cur_lambda;
+    log_cur_lambda = log(cur_lambda);
+    c = 0.767 - 3.36/cur_lambda;
+    beta_rec = sqrt((3./(PI*PI))*cur_lambda);
```

Comment:
That's a micro-optimization to avoid a division. Instead of first multiplying by `3.` and then dividing by `PI`, we simply multiply once and for all by the constant `3./(PI*PI)`, which is (hopefully) pre-computed by the compiler.

Reply:
>  the constant 3./(PI*PI), which is (hopefully) pre-computed by the compiler.
Some C compilers will fold this constant at compile-time, some others may not because its value may depend on the current FP rounding mode...  One way to force compile-time computation is to use a static variable:
```
   static const double three_pi_m2 = 3 ./ (PI*PI);
```